haigha/transports/socket_transport.py

```diff
@@ -29,23 +29,33 @@ def connect(self, (host, port), klass=socket.socket):
         Connect assuming a host and port tuple.
         '''
         self._host = "%s:%s" % (host, port)
-        self._sock = klass()
+        infos = socket.getaddrinfo(host, port, 0, 0, socket.IPPROTO_TCP)
+        if not infos:
+            raise socket.error("getaddrinfo returns an empty list")
+
+        # Sort addresses such that the new IPv6 address is first.
+        for info in reversed(sorted(infos, key=lambda info: len(info[-1]))):
+
+            sockaddr = info[-1]
+            try:
+                # Check if the IP is a valid IPv6 address. Fallback to IPv4 if not.
+                socket.inet_pton(socket.AF_INET6, sockaddr[0])
+                self._sock = klass(socket.AF_INET6)
+                break
+            except socket.error:
+                self._sock = klass()
+                break
+        else:
+
+            raise socket.error("No valid socket object could be created.")
+
         self._sock.setblocking(True)
         self._sock.settimeout(self.connection._connect_timeout)
         if self.connection._sock_opts:
             for k, v in self.connection._sock_opts.iteritems():
                 family, type = k
                 self._sock.setsockopt(family, type, v)
-        infos = socket.getaddrinfo(host, port, 0, 0, socket.IPPROTO_TCP)
-        exc = socket.error("getaddrinfo returns an empty list")
-        for _family, _socktype, _proto, _canonname, sockaddr in infos:
-            try:
-                self._sock.connect(sockaddr)
-                break
-            except socket.error as exc:
-                continue
-        else:
-            raise exc
+        self._sock.connect(sockaddr[:2])
```

Comment:
The `flowinfo` and `scopeid` portions can be omitted in any of the socket functions. It's allowed for backwards compat reasons, but there is a caution in the docs that it may cause problems later on. Next chance I get I'll pluck the slice out and see how it affects the feature.